src/backend/optimizer/plan/createplan.c

```diff
@@ -4394,13 +4394,30 @@ create_ctescan_plan(PlannerInfo *root, Path *best_path,
 		if (!cteplaninfo->shared_plan)
 		{
 			RelOptInfo *sub_final_rel;
+			GangType	saved_gangType = root->curSlice->gangType;
 
 			sub_final_rel = fetch_upper_rel(best_path->parent->subroot, UPPERREL_FINAL, NULL);
 			subplan = create_plan(best_path->parent->subroot, sub_final_rel->cheapest_total_path, root->curSlice);
 			cteplaninfo->shared_plan = prepare_plan_for_sharing(cteroot, subplan);
+
+			/*
+			 * If gangType has switched, it means that CTE's plan contains
+			 * modifying operation without motion above (otherwise, the
+			 * gangType wouldn't switch). In this case we should mark each
+			 * ShareInputScan as writing slice creator, in order to prevent
+			 * the situation, when consumer gets to the writer gang and producer
+			 * gets to the reader gang (it depends of tree traverse order
+			 * inside the apply_shareinput_dag_to_tree function)
+			 */
+			if (root->curSlice->gangType != saved_gangType &&
+				root->curSlice->gangType == GANGTYPE_PRIMARY_WRITER)
+				cteplaninfo->rootSliceIsWriter = true;
```

Comment:
It doesnt work that way. `shared_plan` is Plan node, which is put under ShareInputScan in `share_prepared_plan`. And ShareInputScan is unique for each CTE refernce. In this conditions we need somehow to mark each ShareInputScan corresponding to the same CTE

Reply:
1) `cteplaninfo->shared_plan == NULL`
You do: `cteplaninfo->rootSliceIsWriter = true;`
I suggest: `cteplaninfo->shared_plan->rootSliceIsWriter = true;`
2) next `cteplaninfo->shared_plan != NULL`
You check: `if (cteplaninfo->rootSliceIsWriter)`
I suggest: `if (cteplaninfo->shared_plan->rootSliceIsWriter)`
What is the difference?

Reply:
`cteplaninfo->shared_plan` is a Plan node, which is shared between ShareInputScans, each ShareInputScan is unique, and is created  as ` subplan = share_prepared_plan`. I have to know somehow about the state of other ShareInputScan, but i have common access to `cteplaninfo` or ``cteplaninfo->shared_plan`.
Instead of `cteplaninfo` struct I will have to put `rootSliceIsWriter` to `Plan` struct (and support serializaton for that field).

Reply:
OK, I see